### PR TITLE
Bug 1138782 - validChars() in predictions.js should accept variants and variants of nearBy's

### PR DESCRIPTION
**Pull request by mnjul**

### CODE REVIEW

**Comment by try-server-hook** [comment]
[mnjul Min-Zhong "John" Lu (mnjul)](https://github.com/mnjul) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=6c403aeb6d62)

**Comment by try-server-hook** [comment]
[mnjul Min-Zhong "John" Lu (mnjul)](https://github.com/mnjul) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=f4e0aee77cc6)

**Comment by try-server-hook** [comment]
[mnjul Min-Zhong "John" Lu (mnjul)](https://github.com/mnjul) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=b7b57f14f28d)

File: apps/keyboard/js/imes/latin/predictions.js [hunk 1]
```diff
@@ -418,29 +418,53 @@ var Predictions = function() {
...
-          continue;
-        // If the character does not occur in this language, but there is
-        // a nearby key that does occur, then maybe it is okay
-        if (!nearbyKeys.hasOwnProperty(c))
-          return false;  // no nearby keys, so no suggestions possible
-        var nearby = nearbyKeys[c];
-        for (c in nearby) {
-          if (characterTable.hasOwnProperty(c))
-            continue outer;
+      return Array.from(s).every(function(c) {
```
Comment (janjongboom) [inline apps/keyboard/js/imes/latin/predictions.js]: This seems terribly inefficient. Looping over every character and then for every char we loop over the variant table and then do another indexOf. Can you make a new lookup table (during initialization of the worker or something) where you flatten this data? Then this should be an O(1) operation.

File: apps/keyboard/js/imes/latin/predictions.js [hunk 2]
```diff
@@ -418,29 +418,53 @@ var Predictions = function() {
...
+          return true;
+        }
+
+        if (nearbyKeys.hasOwnProperty(code)) {
+          for (var nearCode in nearbyKeys[code]) {
+            if (characterTable.hasOwnProperty(nearCode)) {
+              return true;
+            }
+
+            var nearChar = String.fromCharCode(nearCode);
```
Comment (janjongboom) [inline apps/keyboard/js/imes/latin/predictions.js]: See above

File: apps/keyboard/js/imes/latin/predictions.js [hunk 3]
```diff
@@ -418,29 +418,53 @@ var Predictions = function() {
...
+        };
+
+
+        // the char is valid if:
+        // - itself is in character table;
+        // - any of its variant forms is in character table;
+        // - some of its nearkey is in character table;
+        // - some of its nearkey's variant forms is in character table;
+
+        if (characterTable.hasOwnProperty(code)) {
```
Comment (janjongboom) [inline apps/keyboard/js/imes/latin/predictions.js]: It feels verbose with all the `hasOwnProperty` but I think it's the style of predictions.js. Rather have `characterTable[code]`, but if it's like this in the rest of the file just leave it.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=58278acf3ecd7e43c50f519140b25907eee99dda

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=add6f7a5c8793cbc277becc85039831229e0ea11

**Comment by autolander** [comment]
http://docs.taskcluster.net/tools/task-graph-inspector/#DYgR-OagTSmJry0FNUORVg

The pull request failed to pass integration tests. It could not be landed, please try again.

**Comment by autolander** [comment]
http://docs.taskcluster.net/tools/task-graph-inspector/#4CV9geYGQJmC-EhHrqojAA

The pull request failed to pass integration tests. It could not be landed, please try again.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=dba9837dadd1afb143ca34dc68dac98b4d6a6889

**Pull request merged**